onnxruntime/core/providers/openvino/backends/basic_backend.h

```diff
@@ -22,16 +28,63 @@ class BasicBackend : public IBackend {
   void Infer(Ort::CustomOpApi& ort, OrtKernelContext* context) override;
 
  private:
-  void StartAsyncInference(Ort::CustomOpApi& ort, OrtKernelContext* context);
+  void StartAsyncInference(Ort::CustomOpApi& ort, OrtKernelContext* context, std::shared_ptr<InferenceEngine::InferRequest> infer_request);
 
-  void CompleteAsyncInference(Ort::CustomOpApi& ort, OrtKernelContext* context);
+  void CompleteAsyncInference(Ort::CustomOpApi& ort, OrtKernelContext* context, std::shared_ptr<InferenceEngine::InferRequest> infer_request);
 
   GlobalContext& global_context_;
   SubGraphContext subgraph_context_;
   mutable std::mutex compute_lock_;
   std::shared_ptr<InferenceEngine::CNNNetwork> ie_cnn_network_;
   std::map<std::string, std::shared_ptr<ngraph::Node>> const_outputs_map_;
-  InferenceEngine::InferRequest::Ptr infer_request_;
+  std::unique_ptr<InferRequestsQueue> inferRequestsQueue_;
 };
+
+class InferRequestsQueue {
+public:
+InferRequestsQueue(InferenceEngine::ExecutableNetwork& net, size_t nireq) {
+  for (size_t id = 0; id < nireq; id++) {
+      infer_request_ = net.CreateInferRequestPtr();
+      infer_requests_.push_back(infer_request_);
+    }
+  }
+
+~InferRequestsQueue() {
+  // clearing out the infer_requests_ vector pool in the class's destructor
+  for(auto& pointer : infer_requests_) {
+       pointer = nullptr;
+   }
+  infer_requests_.erase(std::remove(infer_requests_.begin(), infer_requests_.end(), nullptr), infer_requests_.end());
+}
+
+void printstatus() {
+    std::cout << "printing elements of the vector (infer_requests_): " << std::endl;
+    for (auto i = infer_requests_.begin(); i != infer_requests_.end(); ++i) {
+        std::cout << *i << " ";
+    }
+    std::cout << '\n';
+}
+
+void putIdleRequest(InferenceEngine::InferRequest::Ptr infer_request_) {
+    std::unique_lock<std::mutex> lock(_mutex);
+    infer_requests_.push_back(infer_request_);
+    _cv.notify_one();
+}
+
+InferenceEngine::InferRequest::Ptr getIdleRequest() {
+    std::unique_lock<std::mutex> lock(_mutex);
+    _cv.wait(lock, [this]{ return infer_requests_.size() > 0; });
+    auto request = infer_requests_.at(0);
+    infer_requests_.erase(infer_requests_.begin());
+    return request;
+}
+
+private:
+std::mutex _mutex;
+std::condition_variable _cv;
+std::vector<InferenceEngine::InferRequest::Ptr> infer_requests_;
+InferenceEngine::InferRequest::Ptr infer_request_;
```

Comment:
Could this be removed and replaced with a local variable in the constructor? As far as I can tell, the constructor is the only place where it's read or written; and it's also (somewhat confusingly) shadowed by the argument to `putIdleRequest()`.

Reply:
> Could this be removed and replaced with a local variable in the constructor? As far as I can tell, the constructor is the only place where it's read or written; and it's also (somewhat confusingly) shadowed by the argument to `putIdleRequest()`.

@mrry These instance variables hold values that must be referenced by more than one method from the same class ( putIdleRequest(), printstatus(), getIdleRequest() )

The **Infer()** function call in basic_backend.cc gets called multiple times when we have multiple threads running in parallel for inference. For each Infer() call, we callout all those 3 methods using the **inferRequestsQueue_** object we created.
so these values are read and written multiple times and that is the reason why I had to declare them as instance variables.

Reply:
Perhaps my comment was ambiguous: I'm referring only to the private `InferenceEngine::InferRequest::Ptr infer_request` on this line, which does not appear to be used in the `putIdleRequest()`, `printStatus()` or `getIdleRequest()` methods - all of these use the vector `infer_requests_` instead.

Reply:
> Perhaps my comment was ambiguous: I'm referring only to the private `InferenceEngine::InferRequest::Ptr infer_request` on this line, which does not appear to be used in the `putIdleRequest()`, `printStatus()` or `getIdleRequest()` methods - all of these use the vector `infer_requests_` instead.

Yup, got it now. Makes perfect sense.
I will declare the InferenceEngine::InferRequest::Ptr infer_request as a local variable inside the constructor and push the fix.

Reply:
Thanks!